File: src/main/kotlin/hr/loyalty/program/loyaltyprogramservice/auth/util/JwtUtils.java
package hr.loyalty.program.loyaltyprogramservice.auth.util;

import com.auth0.jwt.JWT;
import com.auth0.jwt.algorithms.Algorithm;
import com.auth0.jwt.interfaces.DecodedJWT;
import com.auth0.jwt.interfaces.JWTVerifier;
import hr.loyalty.program.loyaltyprogramservice.auth.model.User;
import hr.loyalty.program.loyaltyprogramservice.auth.service.UserManagementService;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.stereotype.Component;

@Component
public class JwtUtils {
    private final Algorithm algorithm;
    private final JWTVerifier verifier;
    private final UserManagementService userManagementService;

    public JwtUtils(@Value("{jwt.secret}") String jwtSecret, UserManagementService userManagementService) {
        this.userManagementService = userManagementService;
        this.algorithm = Algorithm.HMAC256(jwtSecret);
        this.verifier = JWT.require(algorithm)
                .withIssuer("auth0")
                .build();
    }

    public String generateToken(Authentication auth) {
        User user = (User) auth.getPrincipal();

        return JWT.create()
                .withIssuer("auth0")
                .withClaim("username", user.getUsername())
                .sign(algorithm);
    }

    public UserDetails getUserFromJWT(String jwt) {
        DecodedJWT decodedJWT = this.decodeJwt(jwt);
        return this.userManagementService.loadUserByUsername(decodedJWT.getClaim("username").asString());
    }

    private DecodedJWT decodeJwt(String jwt) {
        return this.verifier.verify(jwt);
    }
}

File: src/main/kotlin/hr/loyalty/program/loyaltyprogramservice/auth/util/JwtUtils.kt
package hr.loyalty.program.loyaltyprogramservice.auth.util

import com.auth0.jwt.JWT
import com.auth0.jwt.algorithms.Algorithm
import com.auth0.jwt.interfaces.DecodedJWT
import com.auth0.jwt.interfaces.JWTVerifier
import hr.loyalty.program.loyaltyprogramservice.auth.model.User
import org.springframework.beans.factory.annotation.Value
import org.springframework.security.core.Authentication
import org.springframework.stereotype.Component

@Component
class JwtUtils(@Value("{jwt.secret}") private val jwtSecret: String) {
    private val algorithm: Algorithm = Algorithm.HMAC256(jwtSecret)
    private val verifier: JWTVerifier

    init {
        verifier = JWT.require(algorithm)
            .withIssuer("auth0")
            .build()
    }

    fun generateToken(auth: Authentication): String {
        val user = auth.principal as User
        return JWT.create()
            .withIssuer("auth0")
            .withSubject(user.username)
            .sign(algorithm)
    }

    fun decodeJwt(jwt: String): DecodedJWT = verifier.verify(jwt)
}